engine/util/rng.hpp

```diff
@@ -66,26 +244,67 @@ struct basic_rng_t
   double range( double min, double max );
 
   /// Uniform distribution in the range [min..max)
-  template <typename T, typename = std::enable_if_t<std::is_integral<T>::value>>
+  template <typename T, typename = std::enable_if_t<std::numeric_limits<T>::is_integer>>
   T range( T min, T max ) {
     return static_cast<T>(range(static_cast<double>(min), static_cast<double>(max)));
   }
 
   /// Uniform distribution in the range [0..max)
-  template <typename T, typename = std::enable_if_t<std::is_integral<T>::value>>
+  template <typename T, typename = std::enable_if_t<std::numeric_limits<T>::is_integer>>
   T range( T max ) {
     return range<T>( T{}, max );
   }
 
   /// Gaussian Distribution
-  double gauss( double mean, double stddev, bool truncate_low_end = false );
+  double gauss( double mean, double stddev );
+
+  // Truncated Gaussian Distribution
+  double gauss_ab( double mean, double stddev, double min, double max );
+  double gauss_a( double mean, double stddev, double min );
+  double gauss_b( double mean, double stddev, double max );
 
   /// Exponential Distribution
   double exponential( double nu );
 
   /// Exponentially Modified Gaussian Distribution
   double exgauss( double gauss_mean, double gauss_stddev, double exp_nu );
 
+  template <typename T,
+            typename std::enable_if<std::numeric_limits<T>::is_integer && !std::numeric_limits<T>::is_signed, T>::type>
+  T gauss( T mean, T stddev )
+  {
+    return static_cast<T>( gauss( static_cast<double>( mean ), static_cast<double>( stddev ) ) );
+  }
+
+  template <typename T, typename = std::enable_if_t<std::numeric_limits<T>::is_integer>>
+  T gauss_ab( T mean, T stddev, T min, T max )
+  {
+    return static_cast<T>( gauss_ab( static_cast<double>( mean ), static_cast<double>( stddev ),
+                                     static_cast<double>( min ), static_cast<double>( max ) ) );
+  }
+
+  template <typename T, typename = std::enable_if_t<std::numeric_limits<T>::is_integer>>
+  T gauss_a( T mean, T stddev, T min )
+  {
+    return static_cast<T>(
+        gauss_a( static_cast<double>( mean ), static_cast<double>( stddev ), static_cast<double>( min ) ) );
+  }
+
+  template <typename T,
+            typename std::enable_if<std::numeric_limits<T>::is_integer && std::numeric_limits<T>::is_signed, T>::type>
+  T gauss( T mean, T stddev )
+  {
+    return static_cast<T>( gauss_a( static_cast<double>( mean ), static_cast<double>( stddev ),
+                                    static_cast<double>( std::numeric_limits<T>::min() ) ) );
+  }
```

Comment:
I think the `is_signed` checks here and above look backwards. Truncating here should only happen when the distribution is unsigned.

Reply:
I would agree that they are backwards. I have made the appropriate local fix.